File: volleyball_stats/static/js/automatic-fillers.js
// automatic-fillers.js
// Автоматические связи между действиями и метриками (порт из C# AutomaticFillersRulesHolder.cs)

/**
 * InActionAutomaticFiller - автозаполнение метрики внутри одного действия
 * Если leftMetric имеет определённое значение, то автоматически заполняет rightMetric
 * 
 * Пример: Serve.Quality = "-", "/", "!", "+", "#" => Serve.ErrorType = "NoError"
 */
export class InActionAutomaticFiller {
    constructor(actionType, leftMetric, rightMetric, leftValues, rightValue) {
        this.actionType = actionType;        // VolleyActionType (Serve, Reception, etc.)
        this.leftMetric = leftMetric;        // Название левой метрики (например, "Quality")
        this.rightMetric = rightMetric;      // Название правой метрики (например, "ErrorType")
        this.leftValues = leftValues;        // Массив значений левой метрики ["=", "-", "/"]
        this.rightValue = rightValue;        // Значение для правой метрики ("NoError")
    }

    /**
     * Применить правило к действию
     * @param {Object} action - объект действия с metrics: { metricName: value, ... }
     * @returns {boolean} - true если правило применено
     */
    use(action) {
        // Проверяем тип действия
        if (action.actionType !== this.actionType) {
            return false;
        }

        // Проверяем что правая метрика ещё не заполнена
        if (action.metrics && action.metrics[this.rightMetric] !== undefined && action.metrics[this.rightMetric] !== null) {
            return false;
        }

        // Проверяем что левая метрика заполнена
        if (!action.metrics || action.metrics[this.leftMetric] === undefined || action.metrics[this.leftMetric] === null) {
            return false;
        }

        // Получаем значение левой метрики
        const leftValue = action.metrics[this.leftMetric];
        
        // Проверяем что значение левой метрики входит в список разрешённых
        if (this.leftValues.includes(String(leftValue))) {
            // Устанавливаем значение правой метрики
            if (!action.metrics) {
                action.metrics = {};
            }
            action.metrics[this.rightMetric] = this.rightValue;
            return true;
        }

        return false;
    }

    toString() {
        return `${this.actionType}: ${this.leftMetric} (${this.leftValues.join(', ')}) ==> ${this.rightMetric}(${this.rightValue})`;
    }
}

/**
 * SequenceAutomaticFiller - автозаполнение метрики на основе предыдущего действия в сегменте
 * Копирует или устанавливает значение метрики текущего действия на основе предыдущего
 * 
 * Пример: Reception.Quality => Set.ReceptionQuality (копирование)
 * Пример: Set.Combination => Attack.Combination (копирование)
 */
export class SequenceAutomaticFiller {
    constructor(leftActionType, rightActionType, leftMetric, rightMetric, leftValues = null, rightValue = null, isCopying = false) {
        this.leftActionType = leftActionType;    // Тип предыдущего действия (Reception)
        this.rightActionType = rightActionType;  // Тип текущего действия (Set)
        this.leftMetric = leftMetric;            // Метрика предыдущего действия (Quality)
        this.rightMetric = rightMetric;          // Метрика текущего действия (ReceptionQuality)
        this.leftValues = leftValues;            // Массив значений (или null для копирования)
        this.rightValue = rightValue;            // Значение для установки (или null для копирования)
        this.isCopying = isCopying;              // true = копировать значение, false = установить конкретное
    }

    /**
     * Применить правило к действию на основе сегмента
     * @param {Object} currentAction - текущее действие
     * @param {Array} segment - массив предыдущих действий в сегменте
     * @returns {boolean} - true если правило применено
     */
    use(currentAction, segment) {
        // Проверяем что текущее действие соответствует правилу
        if (currentAction.actionType !== this.rightActionType) {
            return false;
        }

        // Ищем в сегменте предыдущее действие нужного типа
        const previousAction = segment.find(action => action.actionType === this.leftActionType);
        
        if (!previousAction) {
            return false;
        }

        // Проверяем что у предыдущего действия есть нужная метрика
        if (!previousAction.metrics || previousAction.metrics[this.leftMetric] === undefined || previousAction.metrics[this.leftMetric] === null) {
            return false;
        }

        const leftValue = previousAction.metrics[this.leftMetric];

        // Режим копирования
        if (this.isCopying) {
            if (!currentAction.metrics) {
                currentAction.metrics = {};
            }
            currentAction.metrics[this.rightMetric] = leftValue;
            return true;
        }

        // Режим условной установки значения
        if (this.leftValues && this.leftValues.includes(String(leftValue))) {
            if (!currentAction.metrics) {
                currentAction.metrics = {};
            }
            currentAction.metrics[this.rightMetric] = this.rightValue;
            return true;
        }

        return false;
    }

    toString() {
        if (this.isCopying) {
            return `${this.leftActionType}: ${this.leftMetric} ==> ${this.rightActionType}: ${this.rightMetric}`;
        }
        return `${this.leftActionType}: ${this.leftMetric} (${this.leftValues.join(', ')}) ==> ${this.rightActionType}: ${this.rightMetric}(${this.rightValue})`;
    }
}

/**
 * AutomaticFillersRulesHolder - хранилище всех правил автозаполнения
 */
export class AutomaticFillersRulesHolder {
    constructor() {
        this.inActionFillers = [];
        this.sequenceFillers = [];
    }

    /**
     * Добавить правило InAction
     */
    addInActionFiller(filler) {
        this.inActionFillers.push(filler);
    }

    /**
     * Добавить правило Sequence
     */
    addSequenceFiller(filler) {
        this.sequenceFillers.push(filler);
    }

    /**
     * Получить все InAction правила для конкретного действия и метрики
     */
    getInActionFillers(actionType, metricType) {
        return this.inActionFillers.filter(
            filler => filler.actionType === actionType && filler.leftMetric === metricType
        );
    }

    /**
     * Получить все Sequence правила для конкретного действия
     */
    getSequenceFillers(rightActionType) {
        return this.sequenceFillers.filter(
            filler => filler.rightActionType === rightActionType
        );
    }

    /**
     * Применить все InAction правила к действию
     */
    applyInActionFillers(action) {
        let applied = false;
        for (const filler of this.inActionFillers) {
            if (filler.use(action)) {
                applied = true;
                console.log(`✅ InAction rule applied: ${filler.toString()}`);
            }
        }
        return applied;
    }

    /**
     * Применить все Sequence правила к действию на основе сегмента
     */
    applySequenceFillers(action, segment) {
        let applied = false;
        for (const filler of this.sequenceFillers) {
            if (filler.use(action, segment)) {
                applied = true;
                console.log(`✅ Sequence rule applied: ${filler.toString()}`);
            }
        }
        return applied;
    }
}

/**
 * Создать и вернуть экземпляр с правилами Kvadratiki
 */
export function createKvadratikiFillers() {
    const holder = new AutomaticFillersRulesHolder();

    // ===== 7 InAction правил из конфига Kvadratiki =====
    
    // 0. Serve: Quality (-, /, !, +, #) => ErrorType (NoError)
    holder.addInActionFiller(new InActionAutomaticFiller(
        'Serve',
        'Quality',
        'ErrorType',
        ['-', '/', '!', '+', '#'],
        'NoError'
    ));

    // 1. Reception: Quality (-, /, !, +, #) => ErrorType (NoError)
    holder.addInActionFiller(new InActionAutomaticFiller(
        'Reception',
        'Quality',
        'ErrorType',
        ['-', '/', '!', '+', '#'],
        'NoError'
    ));

    // 2. Set: Quality (-, /, !, +, #) => ErrorType (NoError)
    holder.addInActionFiller(new InActionAutomaticFiller(
        'Set',
        'Quality',
        'ErrorType',
        ['-', '/', '!', '+', '#'],
        'NoError'
    ));

    // 3. Attack: Quality (-, /, !, +, #) => ErrorType (NoError)
    holder.addInActionFiller(new InActionAutomaticFiller(
        'Attack',
        'Quality',
        'ErrorType',
        ['-', '/', '!', '+', '#'],
        'NoError'
    ));

    // 4. Block: Quality (-, /, !, +, #) => ErrorType (NoError)
    holder.addInActionFiller(new InActionAutomaticFiller(
        'Block',
        'Quality',
        'ErrorType',
        ['-', '/', '!', '+', '#'],
        'NoError'
    ));

    // 5. Defence: Quality (-, /, !, +, #) => ErrorType (NoError)
    holder.addInActionFiller(new InActionAutomaticFiller(
        'Defence',
        'Quality',
        'ErrorType',
        ['-', '/', '!', '+', '#'],
        'NoError'
    ));

    // 6. FreeBall: Quality (-, /, !, +, #) => ErrorType (NoError)
    holder.addInActionFiller(new InActionAutomaticFiller(
        'FreeBall',
        'Quality',
        'ErrorType',
        ['-', '/', '!', '+', '#'],
        'NoError'
    ));

    // ===== 9 Sequence правил из конфига Kvadratiki =====

    // 0. Reception.Quality => Set.ReceptionQuality (копирование)
    holder.addSequenceFiller(new SequenceAutomaticFiller(
        'Reception',      // leftActionType
        'Set',            // rightActionType
        'Quality',        // leftMetric
        'ReceptionQuality', // rightMetric
        null,             // leftValues
        null,             // rightValue
        true              // isCopying = true
    ));

    // 1. Set.Quality => Attack.SetQuality (копирование)
    holder.addSequenceFiller(new SequenceAutomaticFiller(
        'Set',
        'Attack',
        'Quality',
        'SetQuality',
        null,
        null,
        true
    ));

    // 2. Set.Direction => Attack.position (копирование)
    holder.addSequenceFiller(new SequenceAutomaticFiller(
        'Set',
        'Attack',
        'Direction',
        'position',
        null,
        null,
        true
    ));

    // 3. Set.Combination => Attack.Combination (копирование)
    holder.addSequenceFiller(new SequenceAutomaticFiller(
        'Set',
        'Attack',
        'Combination',
        'Combination',
        null,
        null,
        true
    ));

    // 4. Set.BlockersCount => Attack.BlockersCount (копирование)
    holder.addSequenceFiller(new SequenceAutomaticFiller(
        'Set',
        'Attack',
        'BlockersCount',
        'BlockersCount',
        null,
        null,
        true
    ));

    // 5. Attack.Combination => Block.Combination (копирование)
    holder.addSequenceFiller(new SequenceAutomaticFiller(
        'Attack',
        'Block',
        'Combination',
        'Combination',
        null,
        null,
        true
    ));

    // 6. Attack.AttackType => Block.AttackType (копирование)
    holder.addSequenceFiller(new SequenceAutomaticFiller(
        'Attack',
        'Block',
        'AttackType',
        'AttackType',
        null,
        null,
        true
    ));

    // 7. Attack.BlockersCount => Block.BlockersCount (копирование)
    holder.addSequenceFiller(new SequenceAutomaticFiller(
        'Attack',
        'Block',
        'BlockersCount',
        'BlockersCount',
        null,
        null,
        true
    ));

    // 8. Attack.position => Defence.Direction (копирование)
    holder.addSequenceFiller(new SequenceAutomaticFiller(
        'Attack',
        'Defence',
        'position',
        'Direction',
        null,
        null,
        true
    ));

    console.log(`✅ Kvadratiki automatic fillers loaded: ${holder.inActionFillers.length} InAction, ${holder.sequenceFillers.length} Sequence`);
    
    return holder;
}

File: volleyball_stats/app/logic/automatic_fillers.py
"""
Модуль для автоматического заполнения метрик действий
Аналог ActionsLib/ActionTypes/AutomaticFillersRulesHolder.cs
"""
from typing import List, Optional, Dict
from dataclasses import dataclass
import json

from app.models.action import VolleyActionType
from app.models.metric import MetricType
from app.models.sequence import VolleyActionSegment


@dataclass
class InActionAutomaticFiller:
    """
    Автоматический филлер для заполнения метрик внутри одного действия
    Логика: если LeftMetric имеет одно из leftValues, то автоматически установить RightMetric = rightValue
    """
    action_type: VolleyActionType
    left_metric: MetricType
    right_metric: MetricType
    left_values: List[str]  # acceptable values names для left_metric
    right_value: str  # acceptable value name для right_metric
    
    def use(self, player_action_text_repr) -> bool:
        """
        Применяет правило автозаполнения
        Returns: True если метрика была установлена
        """
        # Проверка типа действия
        if player_action_text_repr.action_type != self.action_type:
            return False
        
        # Получаем метрики
        right_metric = player_action_text_repr.get_metric(self.right_metric)
        left_metric = player_action_text_repr.get_metric(self.left_metric)
        
        # Если правая метрика уже заполнена, не перезаписываем
        if right_metric is not None:
            return False
        
        # Если левая метрика не заполнена, ничего не делаем
        if left_metric is None:
            return False
        
        # Проверяем, содержится ли значение левой метрики в списке триггеров
        left_value_name = self.left_metric.acceptable_values[str(left_metric.value)]
        if left_value_name in self.left_values:
            # Устанавливаем правую метрику
            right_obj = self.right_metric.get_object_by_large_name(self.right_value)
            player_action_text_repr.set_metric_by_object(self.right_metric, right_obj)
            return True
        
        return False
    
    def __str__(self) -> str:
        left_vals_str = ", ".join(self.left_values)
        return f"{self.action_type.name}: {self.left_metric.name} ({left_vals_str}) ==> {self.right_metric.name}({self.right_value})"
    
    def to_dict(self) -> Dict:
        """Сериализация в словарь"""
        return {
            "action_type": self.action_type.value,
            "left_metric": self.left_metric.to_dict(),
            "right_metric": self.right_metric.to_dict(),
            "left_values": self.left_values,
            "right_value": self.right_value
        }
    
    @staticmethod
    def from_dict(data: Dict) -> 'InActionAutomaticFiller':
        """Десериализация из словаря"""
        return InActionAutomaticFiller(
            action_type=VolleyActionType(data["action_type"]),
            left_metric=MetricType.from_dict(data["left_metric"]),
            right_metric=MetricType.from_dict(data["right_metric"]),
            left_values=data["left_values"],
            right_value=data["right_value"]
        )


@dataclass
class SequenceAutomaticFiller:
    """
    Автоматический филлер на основе последовательности действий
    Логика: если предыдущее действие (left) имеет определенную метрику,
    то для текущего действия (right) автоматически установить метрику
    """
    left_action_type: VolleyActionType
    right_action_type: VolleyActionType
    left_metric_type: MetricType
    right_metric_type: MetricType
    left_values: Optional[List[str]] = None  # Если None, то копирование значения
    right_value: Optional[str] = None
    is_copying: bool = False  # Если True, копирует значение метрики напрямую
    
    def use(self, current_action_text_repr, segment: VolleyActionSegment) -> bool:
        """
        Применяет правило автозаполнения на основе последовательности
        Returns: True если метрика была установлена
        """
        # Проверяем тип текущего действия
        if current_action_text_repr.action_type != self.right_action_type:
            return False
        
        # Проверяем наличие левого действия в сегменте
        if not segment.contains_action_type(self.left_action_type):
            return False
        
        # Получаем последнее действие нужного типа
        left_action = segment.get_by_action_type(self.left_action_type)
        if left_action is None:
            return False
        
        # Получаем метрику из левого действия
        left_metric = left_action.get_metric(self.left_metric_type)
        if left_metric is None:
            return False
        
        # Режим копирования
        if self.is_copying:
            current_action_text_repr.set_metric_by_object(
                self.right_metric_type,
                left_metric.value
            )
            return True
        
        # Режим условного заполнения
        if self.left_values is not None:
            left_value_name = self.left_metric_type.acceptable_values[str(left_metric.value)]
            if left_value_name in self.left_values:
                right_obj = self.right_metric_type.get_object_by_large_name(self.right_value)
                current_action_text_repr.set_metric_by_object(
                    self.right_metric_type,
                    right_obj
                )
                return True
        
        return False
    
    def __str__(self) -> str:
        if self.is_copying:
            return f"{self.left_action_type.name}: {self.left_metric_type.name} ==> {self.right_action_type.name}: {self.right_metric_type.name}"
        
        left_vals_str = ", ".join(self.left_values) if self.left_values else ""
        return f"{self.left_action_type.name}: {self.left_metric_type.name} ({left_vals_str}) ==> {self.right_action_type.name}: {self.right_metric_type.name}({self.right_value})"
    
    def to_dict(self) -> Dict:
        """Сериализация в словарь"""
        return {
            "left_action_type": self.left_action_type.value,
            "right_action_type": self.right_action_type.value,
            "left_metric_type": self.left_metric_type.to_dict(),
            "right_metric_type": self.right_metric_type.to_dict(),
            "left_values": self.left_values,
            "right_value": self.right_value,
            "is_copying": self.is_copying
        }
    
    @staticmethod
    def from_dict(data: Dict) -> 'SequenceAutomaticFiller':
        """Десериализация из словаря"""
        return SequenceAutomaticFiller(
            left_action_type=VolleyActionType(data["left_action_type"]),
            right_action_type=VolleyActionType(data["right_action_type"]),
            left_metric_type=MetricType.from_dict(data["left_metric_type"]),
            right_metric_type=MetricType.from_dict(data["right_metric_type"]),
            left_values=data.get("left_values"),
            right_value=data.get("right_value"),
            is_copying=data.get("is_copying", False)
        )


class AutomaticFillersRulesHolder:
    """Контейнер для всех правил автоматического заполнения"""
    
    def __init__(self):
        self.in_action_fillers: List[InActionAutomaticFiller] = []
        self.sequence_fillers: List[SequenceAutomaticFiller] = []
    
    def add_in_action_filler(self, filler: InActionAutomaticFiller) -> None:
        """Добавляет InAction filler"""
        self.in_action_fillers.append(filler)
    
    def add_sequence_filler(self, filler: SequenceAutomaticFiller) -> None:
        """Добавляет Sequence filler"""
        self.sequence_fillers.append(filler)
    
    def get_in_action_filler(self, 
                            act_type: VolleyActionType, 
                            metric_type: MetricType) -> List[InActionAutomaticFiller]:
        """Получает все InAction fillers для данного действия и метрики"""
        result = []
        for filler in self.in_action_fillers:
            if filler.action_type == act_type and filler.left_metric == metric_type:
                result.append(filler)
        return result
    
    def get_sequence_fillers(self, right_act: VolleyActionType) -> List[SequenceAutomaticFiller]:
        """Получает все Sequence fillers для данного правого действия"""
        result = []
        for filler in self.sequence_fillers:
            if filler.right_action_type == right_act:
                result.append(filler)
        return result
    
    def to_dict(self) -> Dict:
        """Сериализация в словарь"""
        return {
            "in_action_fillers": [f.to_dict() for f in self.in_action_fillers],
            "sequence_fillers": [f.to_dict() for f in self.sequence_fillers]
        }
    
    @staticmethod
    def from_dict(data: Dict) -> 'AutomaticFillersRulesHolder':
        """Десериализация из словаря"""
        holder = AutomaticFillersRulesHolder()
        
        for filler_data in data.get("in_action_fillers", []):
            holder.add_in_action_filler(InActionAutomaticFiller.from_dict(filler_data))
        
        for filler_data in data.get("sequence_fillers", []):
            holder.add_sequence_filler(SequenceAutomaticFiller.from_dict(filler_data))
        
        return holder
    
    def save_to_file(self, filepath: str) -> None:
        """Сохраняет правила в JSON файл"""
        with open(filepath, 'w', encoding='utf-8') as f:
            json.dump(self.to_dict(), f, ensure_ascii=False, indent=2)
    
    @staticmethod
    def load_from_file(filepath: str) -> 'AutomaticFillersRulesHolder':
        """Загружает правила из JSON файла"""
        with open(filepath, 'r', encoding='utf-8') as f:
            data = json.load(f)
        return AutomaticFillersRulesHolder.from_dict(data)
